### PR TITLE
Add tracing capability in test utilities

### PR DESCRIPTION
**Pull request by levitte**
If a test program goes wrong, it's sometimes helpful to be able to
trace what goes on in libcrypto and libssl.

### CODE REVIEW

File: test/testutil/init.c [hunk 1]
```diff
@@ -7,9 +7,135 @@
...
+    BIO *bio;
+    unsigned int ingroup:1;
+} tracedata;
+
+static size_t internal_trace_cb(const char *buf, size_t cnt,
+                                int category, int cmd, void *vdata)
+{
+    int ret = 0;
+    tracedata *trace_data = vdata;
+    union {
```
Comment (paulidale) [inline test/testutil/init.c]: Ugh.  What about a cast to unsigned long for the one place this is used?
Or are some thread id's not integral?

File: test/testutil/init.c [hunk 2]
```diff
@@ -7,9 +7,135 @@
...
+                 strlen(buffer), buffer);
+        break;
+    case OSSL_TRACE_CTRL_WRITE:
+        ret = BIO_write(trace_data->bio, buf, cnt);
+        break;
+    case OSSL_TRACE_CTRL_END:
+        trace_data->ingroup = 0;
+
+        BIO_ctrl(trace_data->bio, PREFIX_CTRL_SET_PREFIX, 0, NULL);
+
```
Comment (paulidale) [inline test/testutil/init.c]: Excess blank line or deliberate?

File: test/testutil/init.c (hunk 1 above)
Comment (levitte) [reply test/testutil/init.c]: Thread IDs aren't always integral.  You should read the [notes in the Linux man-page for `pthread_self`](http://man7.org/linux/man-pages/man3/pthread_self.3.html#NOTES)

File: test/testutil/init.c (hunk 2 above)
Comment (levitte) [reply test/testutil/init.c]: I'll remove it

File: test/testutil/init.c (hunk 1 above)
Comment (paulidale) [reply test/testutil/init.c]: What about a hexdump of the entire thread_id?
A cast to unsigned long is a kludge :)

Comment (levitte) [reply test/testutil/init.c]: What can I say, it's copied from over here:

https://github.com/openssl/openssl/blob/b720949ec09f4dfbff1becc75ca808997e5b51bd/apps/openssl.c#L128-L170

[sigh] no time for deeper refactoring today, but should probably be part of this PR after all.  I'll put this in WIP for the time being.

**Review by paulidale — APPROVED** [review]

**Comment by levitte** [comment]
Ah, approval, so refactoring will be for another PR...  Okie

**Comment by levitte** [comment]
Merged.

d73458d17ac7e44270a2bcbace45ab99334754b4 Add tracing capability in test utilities

**Comment by paulidale** [comment]
> so refactoring will be for another PR

Yep.  I'll try to get to it tomorrow if you don't.

**Pull request closed without merging**